### DISCUSSION TITLE
Support for React's `renderToPipeableStream` function

### DISCUSSION THREAD

**Discussion by syntaxsorcerer** [remix-run/react-router#10602] [Proposals]
I would like my SSR React application to take advantage of suspension using the `renderToPipeableStream` but it doesn't seem like this library has support for the streaming of HTML and only supports `renderToString`

**Comment by thomaswelton** [comment]
I've been able to get a working implementation of react router with deferred data loaders using a streaming render and suspense.

https://gist.github.com/thomaswelton/e135d2fcf8ee9cce031b3c82d8daa5d9

There should only be two steps

1. In your server side render, update your `StaticRouterProvider` so that you pass the prop `hydrate=false`. By doing this react router will no longer create the `window.staticRouterHydrationData` variable

2. Then we can create this variable ourselves and add support for transferring a server side promise into a client side promise.
If you take a look at the gist above all you should need to do is render my `RouterContext` component during your SSR and pass it the `StaticHandlerContext` provided by React Router. You'll want this component rendered before your main js bundle so that `staticRouterHydrationData` is defined before it's needed by your scripts.

**Reply by wibed** [reply]
@thomaswelton could you make an example because given this function
```ts
...
export default async (req, res, next) => {
  const helmet = Helmet.renderStatic();
  let didError = false;

  const stream = renderToPipeableStream(<App />, {
    onAllReady() {
      res.statusCode = didError ? 500 : 200;
      res.setHeader('Content-type', 'text/html');
      res.write(`<!DOCTYPE html`);
      res.write(`<html ${helmet.htmlAttributes.toString()}>
      <head>
        ${helmet.title.toString()}
        ${helmet.meta.toString()}
        ${helmet.link.toString()}
      </head>
      <body>`);
      res.write(`<div id="root">`);
      stream.pipe(res);
      res.write(`</div>`);
      res.write(
        `<script async data-chunk="main" src="http://localhost:3000/static/main.js"></script>`,
      );
      res.write(`</body></html>`);
    },
...
```

i am having troubles creating the context itself or the usefulness of said code snippet if i could
because i had any object id need to create a static router provider.

see below
```ts
  let staticHandlerContext: StaticHandlerContext = {
    basename: "",
    location: "/",
    statusCode: 200,
    matches: [],
    actionHeaders: {},
    actionData: {},
    loaderHeaders: {},
    loaderData: {},
    activeDeferreds: {},
    errors: {}
  }
et handler = createStaticHandler(routes)
let router = createStaticRouter(handler.dataRoutes, staticHandlerContext)
```


id like to know where i went wrong.
here is the full file: https://termbin.com/hnlu

**Reply by thomaswelton** [reply]
The `StaticHandlerContext` that I use in my code example is the response from the  `staticHandler.query` call
https://reactrouter.com/en/main/routers/create-static-handler#handlerqueryrequest-opts

**Reply by thomaswelton** [reply]
Side note: If you are still using `react-helmet` you're are going to want to read this https://open.nytimes.com/the-future-of-meta-tag-management-for-modern-react-development-ec26a7dc9183

**Reply by brophdawg11** [reply]
I'm curious of the use-cases y'all have for doing DIY/manual SSR with React Router versus using [Remix](https://remix.run/)?  At this point (with the data APIs living in RR), Remix is essentially just a compiler/server-runtime for React Router so you can avoid a lot of these types of manual implementations (streaming, meta, links, etc.) by using the built-in Remix APIs (`defer`, `<Meta>`, `<Links>`).  Is there anything in particular preventing you from moving to Remix?  Maybe it's something we can address!

@thomaswelton hit the nail on the head with the original response here - if you want to stream Promises over the wire in SSR - you need to opt out of hydration and manually serialize and deserialize them.  This is done by the `<Scripts>` tag in Remix when using `defer`.

**Reply by thomaswelton** [reply]
For us it's just that our application predates Remix. Migration from a DIY framework has been discussed, but up until recently there was less of a business case for it. But it's something that we are exploring

**Reply by wibed** [reply]
@thomaswelton i tried following your guidelines, but i am obviously missing the createFetchRequest function somehow.
do you use the official one?
https://reactrouter.com/en/main/guides/ssr ?

if you could provide a small little snippet, oh great one, descent upon this world scarce of resources, for small fracture of your love.

**Reply by wibed** [reply]
figured it out:
```js
import type { Request } from 'express';

export const createFetchHeaders = (requestHeaders?: Request['headers']): Headers => {
	const headers = new Headers()

	for (const [key, values] of Object.entries(requestHeaders ?? {})) {
		if (values) {
			if (Array.isArray(values)) {
				for (const value of values) {
					headers.append(key, value)
				}
			} else {
				headers.set(key, values)
			}
		}
	}

	return headers
}

export const createFetchRequest = (req: Request): globalThis.Request => {
	const origin = `${req.protocol}://${req.get('host')}`
	const url = new URL(req.originalUrl || req.url, origin)

	const controller = new AbortController()

	req.on('close', () => {
		controller.abort()
	})

	const init: RequestInit = {
		method: req.method,
		headers: createFetchHeaders(req.headers),
		signal: controller.signal
	}

	if (req.method !== 'GET' && req.method !== 'HEAD') {
		init.body = req.body
	}

	return new Request(url.href, init)
}
```

**Reply by silvenon** [reply]
@brophdawg11 the primary reason stopping me from migrating my app to Remix is the fact that another service is sitting in front of my app, and uses HTML comments from my app to insert its own markup, which gets erased upon React hydration.

Remix is intended to render entire documents, unless there's a workaround to only render it in part of an HTML template? Until I discover an alternative, I'm using React Router to render a piece of my app into a `<div>`, so that the rest of HTML comments are outside of React and they don't get hydrated.

The secondary reason is waiting for Vite support to stabilize, I need Vite's [Public Base Path](https://vitejs.dev/guide/build.html#public-base-path) feature because my app is on a subpath, and that feature solves my problem perfectly.

**Reply by brophdawg11** [reply]
You can choose to render remix as only a portion of the document but you will give up `<Meta>`/`<Links>` functionality since Remix will no longer be managing the `<head>`.  It just take a few steps:
* Change the `root.tsx` file to return a `<div>` from the component
* `<RemixServer>` will now only render that div's HTML in it's response, so your `entry.server` or your actual server  (Express, etc) will need to stick that in a full document and probably inside a `<div id="app">`
* Update `entry.client` to hydrate the `div#app` instead of the `document`

I don't think this will nicely with streaming features though.  It sounds very error-prone to have react streaming HTML chunks via Suspense along with another downstream apps messing with the HTML.  I would just avoid `defer` on the server while you are in this setup.  You could still implement deferred data on client side navigations via `clientLoader` by returning the raw promise from `serverLoader` and using Suspense/Await since there's no streaming involved there.

Vite `base` support should land soon - keep an eye on https://github.com/remix-run/remix/pull/8145

**Reply by silvenon** [reply]
It felt scary not having Meta and Links because I thought it might screw up Vite or some other important feature, but I assume that Vite's asset insertion is operating outside of those tags.

Thanks for the encouragement and instructions, I'll definitely give it a shot! ❤️ And for the streaming tip, I don't need it and I want to switch over regardless of being able to stream or not.

Regarding the base path, I assume that the asset path of the feature can be offloaded to Vite, so I'll wait for stable Vite plugin support before trying out Remix.
